File: src/previous/day87/MergeSort.ts
export default function merge_sort(arr: number[]): void {
  sort(arr)

  function sort(arr: number[]): void {
    if (arr.length <= 1) return

    let mid = Math.floor(arr.length / 2)
    const left: number[] = []
    const right: number[] = []
    for (let i = 0; i < arr.length; i++) {
      if (i < mid) left.push(arr[i])
      else right.push(arr[i])
    }

    sort(left)
    sort(right)
    merge(arr, left, right)
  }
}
function merge(arr: number[], left: number[], right: number[]): void {
  let [i, l, r] = [0, 0, 0]
  while (l < left.length && r < right.length) {
    if (left[l] < right[r]) arr[i++] = left[l++]
    else if (right[r] <= left[l]) arr[i++] = right[r++]
  }
  while (l < left.length) {
    arr[i++] = left[l++]
  }
  while (r < right.length) {
    arr[i++] = right[r++]
  }
}

File: go/day01/MergeSort/MergeSort.go
package merge_sort

func MergeSort(arr []int) {
	sort(arr)
}

func sort(arr []int) {
	if len(arr) <= 1 {
		return
	}

	mid := len(arr) / 2
	var left []int
	var right []int

	for i := 0; i < len(arr); i++ {
		if i < mid {
			left = append(left, arr[i])
		} else {
			right = append(right, arr[i])
		}
	}

	sort(left)
	sort(right)
	merge(arr, left, right)
}

func merge(arr, left, right []int) {
	i := 0
	l := 0
	r := 0

	for {
		if l < len(left) && r < len(right) {
			if left[l] < right[r] {
				arr[i] = left[l]
				i++
				l++
			} else {
				arr[i] = right[r]
				i++
				r++
			}
		} else if l < len(left) {
			arr[i] = left[l]
			i++
			l++
		} else if r < len(right) {
			arr[i] = right[r]
			i++
			r++
		} else {
			break
		}
	}
}
